File: RadioControl/app/src/main/java/com/nikhilparanjape/radiocontrol/activities/NetworkListActivity.java
package com.nikhilparanjape.radiocontrol.activities;

import android.content.Context;
import android.content.SharedPreferences;
import android.database.DataSetObserver;
import android.graphics.Color;
import android.net.wifi.WifiConfiguration;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SwitchCompat;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.TextView;

import com.mikepenz.google_material_typeface_library.GoogleMaterial;
import com.mikepenz.iconics.IconicsDrawable;
import com.nikhilparanjape.radiocontrol.R;

import java.util.List;

public class NetworkListActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_network_list);
        final ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setHomeAsUpIndicator(new IconicsDrawable(this, GoogleMaterial.Icon.gmd_arrow_back).color(Color.WHITE).sizeDp(IconicsDrawable.TOOLBAR_ICON_SIZE).paddingDp(IconicsDrawable.TOOLBAR_ICON_PADDING));
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setTitle(R.string.title_activity_network);
        }


        ListView List = findViewById(R.id.network_list);
        Thread t = new Thread(() -> {
            listWifiNetworks(List);
        });

        // Start the thread
        t.start();

    }

    private void listWifiNetworks(ListView lv) {
        WifiManager wm = (WifiManager) this.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        final List<WifiConfiguration> networks = wm.getConfiguredNetworks();
        final SharedPreferences prefs = getSharedPreferences("disabled-networks", Context.MODE_PRIVATE);


        ListAdapter la = new ListAdapter() {
            @Override
            public boolean areAllItemsEnabled() {
                return false;
            }

            @Override
            public boolean isEnabled(int position) {
                return false;
            }

            @Override
            public void registerDataSetObserver(DataSetObserver observer) {

            }

            @Override
            public void unregisterDataSetObserver(DataSetObserver observer) {

            }

            @Override
            public int getCount() {
                return networks.size();
            }

            @Override
            public Object getItem(int position) {
                return networks.get(position);
            }

            @Override
            public long getItemId(int position) {
                return ((WifiConfiguration)getItem(position)).networkId;
            }

            @Override
            public boolean hasStableIds() {
                return false;
            }

            @Override
            public View getView(int position, View convertView, ViewGroup parent) {
                if (convertView == null) {


                    LayoutInflater vi = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);

                    WifiConfiguration net = (WifiConfiguration) getItem(position);
                    convertView = vi.inflate(R.layout.wifi_network, parent, false);
                    TextView Name = convertView.findViewById(R.id.net_name);
                    final String ssid = net.SSID.substring(1, net.SSID.length() - 1);
                    Name.setText(ssid);
                    SwitchCompat on = convertView.findViewById(R.id.network_on);
                    Boolean isEnabled = prefs.getBoolean(net.SSID.substring(1, net.SSID.length() - 1), false);
                    on.setChecked(isEnabled);
                    on.setOnCheckedChangeListener((buttonView, isChecked) -> {
                        if(isChecked) {
                            prefs.edit().putBoolean(ssid, true).apply();
                        } else {
                            prefs.edit().remove(ssid).apply();
                        }
                    });


                }



                return convertView;
            }

            @Override
            public int getItemViewType(int position) {
                return 1;
            }

            @Override
            public int getViewTypeCount() {
                return 1;
            }

            @Override
            public boolean isEmpty() {
                return false;
            }
        };
        lv.setAdapter(la);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            // Respond to the action bar's Up/Home button
            case android.R.id.home:
                finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: RadioControl/app/src/main/java/com/nikhilparanjape/radiocontrol/activities/NetworkListActivity.kt
package com.nikhilparanjape.radiocontrol.activities

import android.content.Context
import android.database.DataSetObserver
import android.graphics.Color
import android.net.ConnectivityManager
import android.net.wifi.WifiConfiguration
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ListAdapter
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SwitchCompat
import com.mikepenz.iconics.IconicsDrawable
import com.mikepenz.iconics.typeface.library.googlematerial.GoogleMaterial
import com.mikepenz.iconics.utils.colorInt
import com.mikepenz.iconics.utils.paddingDp
import com.mikepenz.iconics.utils.sizeDp
import com.nikhilparanjape.radiocontrol.R

/**
 * Created by Nikhil on Fill
 *
 * @author
 * An activity that lists all saved WiFi Networks
 */

class NetworkListActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_network_list)
        setSupportActionBar(findViewById(R.id.toolbar))
        val actionBar = supportActionBar

        if (actionBar != null) {
            actionBar.setHomeAsUpIndicator(IconicsDrawable(this, GoogleMaterial.Icon.gmd_arrow_back).apply {
                colorInt = Color.WHITE
                sizeDp = 24
                paddingDp = 1
            })
            actionBar.setDisplayHomeAsUpEnabled(true)
            actionBar.setTitle(R.string.title_activity_network)
        }
        val netList = findViewById<ListView>(R.id.network_list)
        val errorText = findViewById<TextView>(R.id.errorTextView)
        if (Build.VERSION.SDK_INT >= 30){
            //listWifiNetworks will crash on versions above SDK 30 due to deprecations
            // TODO Fix/make an alternative network blocklist solution
            errorText.visibility = View.VISIBLE

        } else{
            listWifiNetworks(netList)
        }

    }

    private fun listWifiNetworks(lv: ListView) {
        //val wm = this.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
        val cm = this.applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        //val networks = wm.configuredNetworks // A deprecated method for listing all configured WiFi networks
        //val networks = wm.connectionInfo // A method for listing the current networks info
        val networks = cm.allNetworks // Use the non-deprecated ConnectivityManager to get a list of all networks
        /*if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED){
            var networks = wm.configuredNetworks
        }*/ // This is no longer needed as WifiManager is deprecated, or at least the configuredNetworks method is.
        val prefs = getSharedPreferences("disabled-networks", Context.MODE_PRIVATE)

        val la = object : ListAdapter {
            override fun areAllItemsEnabled(): Boolean {
                return false
            }

            override fun isEnabled(position: Int): Boolean {
                return false
            }

            override fun registerDataSetObserver(observer: DataSetObserver) {

            }

            override fun unregisterDataSetObserver(observer: DataSetObserver) {

            }

            override fun getCount(): Int {
                return networks.size
            }

            override fun getItem(position: Int): Any {
                return networks[position]
            }

            override fun getItemId(position: Int): Long {
                return (getItem(position) as WifiConfiguration).networkId.toLong()
            }

            override fun hasStableIds(): Boolean {
                return false
            }

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View? {
                var convertViewGet = convertView
                if (convertViewGet == null) {
                    val vi = getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

                    val net = getItem(position) as WifiConfiguration
                    convertViewGet = vi.inflate(R.layout.wifi_network, parent, false)
                    val name = convertViewGet.findViewById<TextView>(R.id.net_name)
                    val ssid = net.SSID.substring(1, net.SSID.length - 1)
                    name.text = ssid
                    val on = convertViewGet.findViewById<SwitchCompat>(R.id.network_on)
                    val isEnabled = prefs.getBoolean(net.SSID.substring(1, net.SSID.length - 1), false)
                    on.isChecked = isEnabled
                    on.setOnCheckedChangeListener { _, isChecked ->
                        if (isChecked) {
                            prefs.edit().putBoolean(ssid, true).apply()
                        } else {
                            prefs.edit().remove(ssid).apply()
                        }
                    }
                }

                return convertViewGet
            }

            override fun getItemViewType(position: Int): Int {
                return 1
            }

            override fun getViewTypeCount(): Int {
                return 1
            }

            override fun isEmpty(): Boolean {
                return false
            }
        }
        lv.adapter = la
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            // Respond to the action bar's Up/Home button
            android.R.id.home -> {
                finish()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

}
